File: bot/src/commands/request.rs
use chrono::{NaiveDateTime, Utc};
use diesel::{
    insert_into, update, BelongingToDsl, ExpressionMethods, PgConnection, QueryDsl, RunQueryDsl,
};
use mlua::{Lua, Table};
use substring::Substring;
use twitch_irc::message::PrivmsgMessage;

use common::{
    config::CommandsConfiguration,
    models::{
        Channel, ChannelPreference, LevelOfRights, NewChannel, NewChannelPreference, NewRight,
        NewUser, Right, User,
    },
    schema::{
        actions::dsl as ac, channel_preferences::dsl as chp, channels::dsl as ch,
        rights::dsl as ri, users::dsl as us,
    },
};

use super::CommandLoader;

#[derive(Clone)]
pub struct Request {
    pub command_id: String,
    pub subcommand_id: Option<String>,
    pub message: Option<String>,

    pub sender: User,
    pub channel: Channel,
    pub channel_preference: ChannelPreference,
    pub rights: Right,
}

impl Request {
    pub fn try_from(
        config: &CommandsConfiguration,
        message: &PrivmsgMessage,
        command_loader: &CommandLoader,
        conn: &mut PgConnection,
    ) -> Option<Request> {
        let channel = ch::channels
            .filter(ch::alias_id.eq(message.channel_id.parse::<i32>().unwrap()))
            .first::<Channel>(conn)
            .unwrap_or_else(|_| {
                insert_into(ch::channels)
                    .values(vec![NewChannel {
                        alias_id: message.channel_id.parse::<i32>().unwrap(),
                        alias_name: message.channel_login.clone(),
                    }])
                    .execute(conn)
                    .expect("Failed to create a new channel");

                ch::channels
                    .filter(ch::alias_id.eq(message.channel_id.parse::<i32>().unwrap()))
                    .first::<Channel>(conn)
                    .expect("Failed to get a channel after creating it")
            });

        let channel_preference = ChannelPreference::belonging_to(&channel)
            .first::<ChannelPreference>(conn)
            .unwrap_or_else(|_| {
                insert_into(chp::channel_preferences)
                    .values(vec![NewChannelPreference {
                        channel_id: channel.id,
                        prefix: config.default_prefix.clone(),
                        language: config.default_language.clone(),
                    }])
                    .execute(conn)
                    .expect("Failed to create preferences for channel");

                ChannelPreference::belonging_to(&channel)
                    .first::<ChannelPreference>(conn)
                    .expect("Failed to get preferences after creating them")
            });

        let mut sender = us::users
            .filter(us::alias_id.eq(message.sender.id.parse::<i32>().unwrap()))
            .first::<User>(conn)
            .unwrap_or_else(|_| {
                insert_into(us::users)
                    .values(vec![NewUser {
                        alias_id: message.sender.id.parse::<i32>().unwrap(),
                        alias_name: message.sender.login.clone(),
                    }])
                    .execute(conn)
                    .expect("Failed to create a new user");

                us::users
                    .filter(us::alias_id.eq(message.sender.id.parse::<i32>().unwrap()))
                    .first::<User>(conn)
                    .expect("Failed to get a user after creating it")
            });

        if sender.alias_name.ne(&message.sender.login) {
            sender.alias_name = message.sender.login.clone();

            update(us::users.find(sender.id))
                .set(us::alias_name.eq(&message.sender.login))
                .execute(conn)
                .expect("Failed to update user name");
        }

        let mut badges_iter = message.badges.iter();

        let level_of_rights = if sender.alias_id == channel.alias_id {
            LevelOfRights::Broadcaster
        } else if badges_iter.any(|x| x.name.eq("moderator")) {
            LevelOfRights::Moderator
        } else if badges_iter.any(|x| x.name.eq("vip")) {
            LevelOfRights::Vip
        } else if badges_iter.any(|x| x.name.eq("subscriber")) {
            LevelOfRights::Subscriber
        } else {
            LevelOfRights::User
        };

        let mut rights = ri::rights
            .filter(ri::user_id.eq(&sender.id))
            .get_result::<Right>(conn)
            .unwrap_or_else(|_| {
                insert_into(ri::rights)
                    .values([NewRight {
                        user_id: sender.id,
                        channel_id: channel.id,
                        level: level_of_rights.clone(),
                    }])
                    .get_result::<Right>(conn)
                    .expect("Failed to insert a new rights")
            });

        if rights.level != LevelOfRights::Suspended && rights.level != level_of_rights {
            rights.level = level_of_rights.clone();

            update(ri::rights.find(&rights.id))
                .set(ri::level.eq(level_of_rights))
                .execute(conn)
                .expect("Failed to update rights");
        }

        let prefix = channel_preference.prefix.as_str();

        if !message.message_text.starts_with(prefix) {
            return None;
        }

        let message = message
            .message_text
            .substring(prefix.len(), message.message_text.len())
            .to_string();
        let mut message_split = message.split(' ').collect::<Vec<&str>>();

        let command_id = if let Some(word) = message_split.first() {
            let word = word.to_string();
            if command_loader
                .rust_commands
                .iter()
                .any(|x| x.get_name().eq(&word))
                || command_loader.lua_commands.iter().any(|x| x.name.eq(&word))
            {
                word
            } else {
                return None;
            }
        } else {
            return None;
        };

        let command_rights: LevelOfRights;
        let command_delay: i32;
        let command_subcommands: Vec<String>;

        match (
            command_loader
                .rust_commands
                .iter()
                .find(|x| x.get_name().eq(&command_id)),
            command_loader
                .lua_commands
                .iter()
                .find(|x| x.name.eq(&command_id)),
        ) {
            (Some(c), _) => {
                command_rights = c.required_rights().clone();
                command_delay = c.get_delay_sec();
                command_subcommands = c.get_subcommands().clone();
            }
            (_, Some(c)) => {
                command_rights = c.minimal_rights.clone();
                command_delay = c.delay_sec as i32;
                command_subcommands = c.subcommands.clone();
            }
            _ => return None,
        }

        if command_rights > rights.level {
            return None;
        }

        let last_action_timestamp = ac::actions
            .filter(ac::channel_id.eq(&channel.id))
            .filter(ac::user_id.eq(&sender.id))
            .filter(ac::command_name.eq(&command_id))
            .select(ac::processed_at)
            .order(ac::processed_at.desc())
            .first::<NaiveDateTime>(conn);

        if let Ok(last_action_timestamp) = last_action_timestamp {
            let la_timestamp: i64 = last_action_timestamp.timestamp();
            let now_timestamp: i64 = Utc::now().naive_utc().timestamp();

            if now_timestamp - la_timestamp < command_delay as i64 {
                return None;
            }
        }

        message_split.remove(0);

        let subcommand_id = if let Some(v) = message_split.first() {
            let v = v.to_string();

            if command_subcommands.contains(&v) {
                message_split.remove(0);
                Some(v)
            } else {
                None
            }
        } else {
            None
        };

        Some(Request {
            command_id,
            subcommand_id,
            message: if message_split.is_empty() {
                None
            } else {
                Some(message_split.join(" "))
            },
            sender,
            channel,
            channel_preference,
            rights,
        })
    }

    pub fn as_lua_table(&self, lua: &Lua) -> mlua::Result<Table> {
        let table = lua.create_table()?;

        // metadata
        table.set("command_id", self.command_id.clone())?;
        table.set("subcommand_id", self.subcommand_id.clone())?;
        table.set("message", self.message.clone())?;

        // sender
        let sender_table = lua.create_table()?;
        sender_table.set("id", self.sender.id)?;
        sender_table.set("alias_id", self.sender.alias_id)?;
        sender_table.set("alias_name", self.sender.alias_name.clone())?;
        sender_table.set("joined_at", self.sender.joined_at.timestamp_millis())?;
        sender_table.set(
            "opted_out_at",
            if let Some(datetime) = &self.sender.opt_outed_at {
                Some(datetime.timestamp_millis())
            } else {
                None
            },
        )?;
        table.set("sender", sender_table)?;

        // channel
        let channel_table = lua.create_table()?;
        channel_table.set("id", self.channel.id)?;
        channel_table.set("alias_id", self.channel.alias_id)?;
        channel_table.set("alias_name", self.channel.alias_name.clone())?;
        channel_table.set("joined_at", self.channel.joined_at.timestamp_millis())?;
        channel_table.set(
            "opted_out_at",
            if let Some(datetime) = &self.channel.opt_outed_at {
                Some(datetime.timestamp_millis())
            } else {
                None
            },
        )?;
        table.set("channel", channel_table)?;

        // channel preference
        let channel_preference_table = lua.create_table()?;
        channel_preference_table.set("id", self.channel_preference.id)?;
        channel_preference_table.set("channel_id", self.channel_preference.channel_id)?;
        channel_preference_table.set("prefix", self.channel_preference.prefix.clone())?;
        channel_preference_table.set("language", self.channel_preference.language.clone())?;
        channel_preference_table.set("features", self.channel_preference.features.clone())?;
        table.set("channel_preference", channel_preference_table)?;

        // rights
        let rights_table = lua.create_table()?;
        rights_table.set("id", self.rights.id)?;
        rights_table.set("user_id", self.rights.user_id)?;
        rights_table.set("channel_id", self.rights.channel_id)?;
        rights_table.set("level", self.rights.level.to_string())?;
        rights_table.set("is_fixed", self.rights.is_fixed)?;
        table.set("rights", rights_table)?;

        Ok(table)
    }

    pub fn from_lua_table(table: Table) -> mlua::Result<Self> {
        // metadata
        let command_id: String = table.get("command_id")?;
        let subcommand_id: Option<String> = table.get("subcommand_id")?;
        let message: Option<String> = table.get("message")?;

        // sender
        let sender: Table = table.get("sender")?;
        let sender = User {
            id: sender.get("id")?,
            alias_id: sender.get("alias_id")?,
            alias_name: sender.get("alias_name")?,
            joined_at: NaiveDateTime::from_timestamp_millis(sender.get("joined_at")?)
                .expect("Error parsing NaiveDateTime"),
            opt_outed_at: if let Ok(timestamp) = sender.get::<i64>("opted_out_at") {
                Some(
                    NaiveDateTime::from_timestamp_millis(timestamp)
                        .expect("Error parsing NaiveDateTime"),
                )
            } else {
                None
            },
        };

        // channel
        let channel: Table = table.get("channel")?;
        let channel = Channel {
            id: channel.get("id")?,
            alias_id: channel.get("alias_id")?,
            alias_name: channel.get("alias_name")?,
            joined_at: NaiveDateTime::from_timestamp_millis(channel.get("joined_at")?)
                .expect("Error parsing NaiveDateTime"),
            opt_outed_at: if let Ok(timestamp) = channel.get::<i64>("opted_out_at") {
                Some(
                    NaiveDateTime::from_timestamp_millis(timestamp)
                        .expect("Error parsing NaiveDateTime"),
                )
            } else {
                None
            },
        };

        // channel preference
        let channel_preference: Table = table.get("channel_preference")?;
        let channel_preference = ChannelPreference {
            id: channel_preference.get("id")?,
            channel_id: channel_preference.get("channel_id")?,
            prefix: channel_preference.get("prefix")?,
            language: channel_preference.get("language")?,
            features: channel_preference
                .get::<Table>("features")?
                .sequence_values::<String>()
                .map(|x| x.ok())
                .collect::<Vec<Option<String>>>(),
        };

        // rights
        let rights: Table = table.get("rights")?;
        let rights = Right {
            id: rights.get("id")?,
            user_id: rights.get("user_id")?,
            channel_id: rights.get("channel_id")?,
            level: LevelOfRights::from_str(&rights.get::<String>("level")?),
            is_fixed: rights.get("is_fixed")?,
        };

        Ok(Request {
            command_id,
            subcommand_id,
            message,
            sender,
            channel,
            channel_preference,
            rights,
        })
    }
}

File: bot/src/commands/request.cpp
#include "commands/request.hpp"

#include <algorithm>
#include <optional>
#include <sol/types.hpp>
#include <string>
#include <vector>

#include "config.hpp"
#include "constants.hpp"
#include "database.hpp"
#include "schemas/channel.hpp"
#include "schemas/user.hpp"
#include "utils/string.hpp"

namespace bot::command {
  sol::table Request::as_lua_table(std::shared_ptr<sol::state> luaState) const {
    sol::table o = luaState->create_table();

    o["command_id"] = this->command_id;
    if (this->subcommand_id.has_value()) {
      o["subcommand_id"] = this->subcommand_id.value();
    } else {
      o["subcommand_id"] = sol::lua_nil;
    }
    if (this->message.has_value()) {
      o["message"] = this->message.value();
    } else {
      o["message"] = sol::lua_nil;
    }

    o["sender"] = requester.user.as_lua_table(luaState);
    o["channel"] = requester.channel.as_lua_table(luaState);
    o["channel_preference"] =
        requester.channel_preferences.as_lua_table(luaState);
    o["rights"] = requester.user_rights.as_lua_table(luaState);

    return o;
  }

  std::optional<Requester> get_requester(
      const irc::Message<irc::MessageType::Privmsg> &irc_message,
      std::unique_ptr<db::BaseDatabase> &conn, const Configuration &cfg) {
    // fetching channel
    std::vector<schemas::Channel> chans = conn->query_all<schemas::Channel>(
        "SELECT * FROM channels WHERE alias_id = $1",
        {std::to_string(irc_message.source.id)});

    if (chans.empty()) {
      conn->exec(
          "INSERT INTO channels(alias_id, alias_name) VALUES ($1, $2)",
          {std::to_string(irc_message.source.id), irc_message.source.login});

      chans = conn->query_all<schemas::Channel>(
          "SELECT * FROM channels WHERE alias_id = $1",
          {std::to_string(irc_message.source.id)});
    }

    schemas::Channel channel = chans[0];
    if (channel.get_opted_out_at().has_value()) {
      return std::nullopt;
    }

    // fetching channel preference
    std::vector<schemas::ChannelPreferences> prefs =
        conn->query_all<schemas::ChannelPreferences>(
            "SELECT * FROM channel_preferences WHERE id = $1",
            {std::to_string(channel.get_id())});

    if (prefs.empty()) {
      conn->exec(
          "INSERT INTO channel_preferences(id, prefix, locale) VALUES ($1, "
          "$2, $3)",
          {std::to_string(channel.get_id()), DEFAULT_PREFIX,
           DEFAULT_LOCALE_ID});

      prefs = conn->query_all<schemas::ChannelPreferences>(
          "SELECT * FROM channel_preferences WHERE id = $1",
          {std::to_string(channel.get_id())});
    }

    schemas::ChannelPreferences pref = prefs[0];

    // fetching channel preference
    std::vector<schemas::User> users = conn->query_all<schemas::User>(
        "SELECT * FROM users WHERE alias_id = $1",
        {std::to_string(irc_message.sender.id)});

    if (users.empty()) {
      conn->exec(
          "INSERT INTO users(alias_id, alias_name) VALUES ($1, "
          "$2)",
          {std::to_string(irc_message.sender.id), irc_message.sender.login});

      users = conn->query_all<schemas::User>(
          "SELECT * FROM users WHERE alias_id = $1",
          {std::to_string(irc_message.sender.id)});
    }

    schemas::User user = users[0];

    // updating username
    if (user.get_alias_name() != irc_message.sender.login) {
      conn->exec("UPDATE users SET alias_name = $1 WHERE id = $2",
                 {irc_message.sender.login, std::to_string(user.get_id())});

      user.set_alias_name(irc_message.sender.login);
    }

    // setting permissions
    schemas::PermissionLevel level = schemas::PermissionLevel::USER;
    const auto &badges = irc_message.sender.badges;

    if (std::any_of(
            cfg.twitch.superuser_ids.begin(), cfg.twitch.superuser_ids.end(),
            [&user](const int &x) { return x == user.get_alias_id(); })) {
      level = schemas::PermissionLevel::SUPERUSER;
    } else if (std::any_of(cfg.twitch.trusted_user_ids.begin(),
                           cfg.twitch.trusted_user_ids.end(),
                           [&user](const int &x) {
                             return x == user.get_alias_id();
                           })) {
      level = schemas::PermissionLevel::TRUSTED;
    } else if (user.get_alias_id() == channel.get_alias_id()) {
      level = schemas::PermissionLevel::BROADCASTER;
    } else if (std::any_of(badges.begin(), badges.end(), [&](const auto &x) {
                 return x.first == "moderator";
               })) {
      level = schemas::PermissionLevel::MODERATOR;
    } else if (std::any_of(badges.begin(), badges.end(),
                           [&](const auto &x) { return x.first == "vip"; })) {
      level = schemas::PermissionLevel::VIP;
    }

    std::vector<schemas::UserRights> user_rights =
        conn->query_all<schemas::UserRights>(
            "SELECT * FROM user_rights WHERE user_id = $1 AND channel_id = $2",
            {std::to_string(user.get_id()), std::to_string(channel.get_id())});

    if (user_rights.empty()) {
      conn->exec(
          "INSERT INTO user_rights(user_id, channel_id, level) VALUES ($1, "
          "$2, $3)",
          {std::to_string(user.get_id()), std::to_string(channel.get_id()),
           std::to_string(level)});

      user_rights = conn->query_all<schemas::UserRights>(
          "SELECT * FROM user_rights WHERE user_id = $1 AND channel_id = $2",
          {std::to_string(user.get_id()), std::to_string(channel.get_id())});
    }

    schemas::UserRights user_right = user_rights[0];

    if (user_right.get_level() != level) {
      conn->exec("UPDATE user_rights SET level = $1 WHERE id = $2",
                 {std::to_string(level), std::to_string(user_right.get_id())});

      user_right.set_level(level);
    }

    return (Requester){channel, pref, user, user_right};
  }

  std::optional<Request> generate_request(
      const command::CommandLoader &command_loader,
      const irc::Message<irc::MessageType::Privmsg> &irc_message,
      const Requester &requester, std::unique_ptr<db::BaseDatabase> &conn) {
    // --- FETCHING MESSAGES
    std::string fullmsg = irc_message.message;
    const std::string &prefix = requester.channel_preferences.get_prefix();

    if (fullmsg.empty() || fullmsg.substr(0, prefix.length()) != prefix ||
        std::any_of(requester.channel_preferences.get_features().begin(),
                    requester.channel_preferences.get_features().end(),
                    [](const schemas::ChannelFeature &f) {
                      return f == schemas::ChannelFeature::SILENT_MODE;
                    })) {
      return std::nullopt;
    }

    fullmsg = fullmsg.substr(prefix.length());

    std::vector<std::string> parts = utils::string::split_text(fullmsg, ' ');

    std::string command_id = parts[0];

    auto &cmds = command_loader.get_commands();
    auto cmd =
        std::find_if(cmds.begin(), cmds.end(), [&command_id](const auto &c) {
          auto aliases = c->get_aliases();
          return c->get_name() == command_id ||
                 std::any_of(aliases.begin(), aliases.end(),
                             [&command_id](const std::string &alias) {
                               return alias == command_id;
                             });
        });

    if (cmd == cmds.end()) {
      return std::nullopt;
    }

    parts.erase(parts.begin());

    command_id = (*cmd)->get_name();

    Request req{command_id, std::nullopt, std::nullopt, irc_message, requester};

    if (parts.empty()) {
      return req;
    }

    std::optional<std::string> scid = parts[0];
    auto scids = (*cmd)->get_subcommand_ids();

    if (std::any_of(scids.begin(), scids.end(),
                    [&](const auto &x) { return x == scid.value(); })) {
      parts.erase(parts.begin());
    } else {
      scid = std::nullopt;
    }

    req.subcommand_id = scid;

    std::optional<std::string> message = utils::string::join_vector(parts, ' ');
    req.message = message;

    return req;
  }
}